src/scenic/syntax/compiler.py

```diff
@@ -253,6 +274,94 @@ def visit_Require(self, node: s.Require):
             )
         )
 
+    def visit_Take(self, node: s.Take):
+        action = ast.Tuple(self.visit(node.elts), loadCtx)
+        return self.generateInvocation(node, ast.copy_location(action, node))
+
+    def visit_Wait(self, node: s.Wait):
+        return self.generateInvocation(node, ast.Constant(()))
+
+    def visit_Terminate(self, node: s.Terminate):
+        termination = ast.Call(
+            ast.Name("makeTerminationAction", loadCtx), [ast.Constant(node.lineno)], []
+        )
+        return self.generateInvocation(node, termination)
+
+    def visit_Do(self, node: s.Do):
+        subHandler = ast.Attribute(
```

Comment:
`visit_Do` `visit_DoFor` and `visit_DoUntil` are similar and might be able to extract the common code to a helper function. I want to wait until I implement `do choose` and `do shuffle` so I can choose the right level of abstraction for that function.